File: 30 days of python/2024/Algo/array.py
#initialize an array 
arr=[1,2,3,4,5]
# time comlexity is 0(1) 
# Iterate through an array 
for elem in arr:
    print(elem)
# 0(n) complexity you visit all elements once 
# Get element by index 0(1) 
elem = arr[2]

# search an elemet 0(n) complexity 

if 3 in arr:
    print("Found 3")
# Insert elemt 0(n)
arr.insert(3,5 )

#delete an element
del arr[2]
#filtered array 
filterered_arr = [x for x in arr if x >= 3]

#fetch subarray 
sub_array = arr[1:4]
# merege array - complexity 0(n+m) merging two requires visiting each element in both arrays,where n and m are the sizes of the arrays
arr1 = [1,2,3]
arr2= [4,5.6]
merge_arr = arr1 + arr2

# Reverse an array 0(n) requires visiting each element once and swapping elements,which takes linear time
reversed_arr = arr[::-1] # revers the array

# rotate an array 0(n)
rotate_by = 2
rotated_arr =  arr[rotate_by:] + arr[:rotate_by] # rotate array by 2

File: 30 days of python/2024/Algo/array.go
//
arr := []int{1,2,3,4,5}

// iterate through an array 
for _, elem := range arr{
	fmt.Println(elem)
}
//get element from slice
elem := arr[2]

//check element in an array 0(n) you may need to loop through the whole list
found  := false
for _, elem= range arr{
    if  elem == 3 {
	    found = true
		break
	}
}
if found {
	fmt.Println("found 3")
}
// insert an elemet 
arr = append(arr[:2], append([]int{10},arr[2:]...)...)

//delete an item
arr = append(arr[:2], arr[3:]) //Deletes elemet at index 2

//filter an array creatin an array that only contains elements that meet a certain condition
// time complexity 0(n)

var filterArr []int
for _, := range arr {
	if x >3 {
		filterArr = append(filter, x)
	}
}

//fetch subarray extract a section of array from one index to another
// complexity 0(k) - creating a subarray where k is the size of the sub-array you're fetching
 subArr := arr[1:4] //fetches subarray from index 1 to 3

 //merge array 
 arr1 := []int{1,2,3}
 arr2 := []int{4,5,6}
 mergeArr := append(arr1,arr2)

 // reverse array 
 for i,j :=0, len (arr)-1; i <j ; i,j = i+1, j-1 {
	arr[1],arr[j] = arr[j],arr[i]
 }

 // rotate array means shifting elements to the left or right 
 rotateBy := 2
 rotatedArr =: append(arr[rotateBy:], arr[:rotateBy])